src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,562 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.marshal;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.cassandra.cql3.CQL3Type;
+import org.apache.cassandra.cql3.Term;
+import org.apache.cassandra.cql3.Vectors;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.serializers.TypeSerializer;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.JsonUtils;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+public final class VectorType<T> extends AbstractType<List<T>>
+{
+    public final AbstractType<T> elementType;
+    public final int dimention;
+    protected final TypeSerializer<T> elementSerializer;
+    private final int valueLengthIfFixed;
+    private final VectorSerializer serializer;
+
+    private VectorType(AbstractType<T> elementType, int dimention)
+    {
+        super(ComparisonType.CUSTOM);
+        if (dimention <= 0)
+            throw new InvalidRequestException(String.format("vectors may only have positive dimentions; given %d", dimention));
+        this.elementType = elementType;
+        this.dimention = dimention;
+        this.elementSerializer = elementType.getSerializer();
+        this.valueLengthIfFixed = elementType.isValueLengthFixed() ?
+                                  elementType.valueLengthIfFixed() * dimention :
+                                  super.valueLengthIfFixed();
+        this.serializer = elementType.isValueLengthFixed() ?
+                          new FixedLengthSerializer() :
+                          new VariableLengthSerializer();
+    }
+
+    public static <T> VectorType<T> getInstance(AbstractType<T> elements, int dimention)
+    {
+        return new VectorType<>(elements, dimention);
```

Comment:
TODO cache like the others?

Reply:
I think we definitely need a cache for these types. They are created in a number of places.

Reply:
created cache...